### PR TITLE
Kernel: Add support for the FAT32 filesystem

### PR DESCRIPTION
**Pull request by Undefine-Unity**
This commit adds read-only support for the FAT32 filesystem. It also includes support for long file names. As far as I know all the FAT32 and long file names patents have already expired, so we should be safe from legal issues. In the future I'm planning to implement write support and FAT12/FAT16 support, although I don't know if we even need it. Also FAT32 will be useful when it comes to UEFI booting, RPi booting and removable devices.

### CODE REVIEW

**Comment by supercomputer7** [comment]
As long as we won't have legal issues (I am not a lawyer) with FAT32 patents, I'm OK with this. I know in fact that exFAT requires a license from Microsoft to be implemented even in open source projects, but I guess that you did your checks to confirm that FAT32 will not be an issue for us.

**Comment by filiphsps** [comment]
> As long as we won't have legal issues (I am not a lawyer) with FAT32 patents, I'm OK with this. I know in fact that exFAT requires a license from Microsoft to be implemented even in open source projects, but I guess that you did your checks to confirm that FAT32 will not be an issue for us.

From what I gather by reading patents we're still not allowed to implement exFAT without a license, see [US20080168029](https://patents.google.com/patent/US20080168029).
vFAT aka FAT32 with long file names expired in 2013 in the states (see [US20080168029](https://patents.google.com/patent/US20080168029)) and 2014 in Germany (see [EP0618540](https://data.epo.org/gpi/EP0618540B1-Common-name-space-for-long-and-short-filenames)).

vFAT is still however debated and IIRC the legal workaround that the likes of Linux uses is to read long file names separately thus circumventing the patent(s).

TL;DR FAT32 in itself is free to use, it's the extensions that are questionable.

(I am not a lawyer, and this is not legal advice. Please don't sue me).

**Comment by Undefine-Unity** [comment]
First of all the EFI FAT32 specification gives permission for implementing FAT32 including long file names for purposes of EFI booting, which we will eventually use the driver for. Also all the relevant long file name / vFAT patents (see [US5579517](https://www.google.com/patents/US5579517), [US5745902](https://www.google.com/patents/US5745902), [US5758352](https://www.google.com/patents/US5758352) and [US6286013](https://www.google.com/patents/US6286013)) seem to have already expired.

File: Kernel/FileSystem/FATFileSystem.cpp [hunk 1]
```diff
@@ -0,0 +1,361 @@
...
+}
+
+Inode& FATFS::root_inode()
+{
+    return *m_root_inode;
+}
+
+BlockBasedFileSystem::BlockIndex FATFS::first_block_of_cluster(u32 cluster) const
+{
+    return ((cluster - 2) * boot_record()->sectors_per_cluster) + m_first_data_sector;
```
Comment (Quaker762) [inline Kernel/FileSystem/FATFileSystem.cpp:82]: Could you be a dear and make this a constant please :^)

File: Kernel/FileSystem/FATFileSystem.cpp [hunk 2]
```diff
@@ -0,0 +1,361 @@
...
+
+ErrorOr<Vector<BlockBasedFileSystem::BlockIndex>> FATInode::compute_block_list()
+{
+    VERIFY(m_inode_lock.is_locked());
+
+    dbgln_if(FAT_DEBUG, "FATFS: computing block list for inode {}", index());
+
+    u32 cluster = (((u32)m_entry.first_cluster_high) << 16) | m_entry.first_cluster_low;
+
+    if (cluster == 1)
```
Comment (Quaker762) [inline Kernel/FileSystem/FATFileSystem.cpp:124]: Same here

File: Kernel/FileSystem/FATFileSystem.cpp [hunk 3]
```diff
@@ -0,0 +1,361 @@
...
+        length -= 1;
+    }
+    return StringView(string, max_length);
+}
+
+ErrorOr<size_t> FATInode::read_bytes_locked(off_t offset, size_t size, UserOrKernelBuffer& buffer, OpenFileDescription*) const
+{
+    dbgln_if(FAT_DEBUG, "FATFS: Reading inode {}: ize: {} offset: {}", identifier().index(), size, offset);
+
+    // FIXME: Read only the neaded blocks instead of the whole file
```
Comment (Quaker762) [inline Kernel/FileSystem/FATFileSystem.cpp:281]: Typo: "neaded"->"needed"

File: Kernel/FileSystem/FATFileSystem.cpp [hunk 4]
```diff
@@ -0,0 +1,361 @@
...
+ErrorOr<NonnullLockRefPtr<Inode>> FATInode::lookup(StringView name)
+{
+    MutexLocker locker(m_inode_lock);
+
+    VERIFY(m_entry.attributes & FATAttributes::Directory);
+
+    Vector<RefPtr<FATInode>> children = TRY(traverse());
+
+    for (auto& inode : children) {
+        // FIXME: For some reason using == doesn't work
```
Comment (Quaker762) [inline Kernel/FileSystem/FATFileSystem.cpp:317]: What are you trying to use `operator==` on? The `inode` specifically? It doesn't look like you have `operator==()` implemented in your `FATInode` class, did you perhaps mean to implement that?

**Review by Quaker762 — CHANGES_REQUESTED** [review]
Looks pretty good, few nits/questions :^)

File: Kernel/FileSystem/FATFileSystem.cpp (hunk 4 above)
Comment (Undefine-Unity) [reply Kernel/FileSystem/FATFileSystem.cpp:317]: I tried comparing `inode->m_filename` and `name`, although it would always return false, later I found out that this is caused by the fact that `KString` adds a null byte at the end of the string. The thing is that using `strncmp` works just fine so I'm thinking of leaving this as it is.

File: Kernel/FileSystem/FATFileSystem.cpp (hunk 1 above)
Comment (Undefine-Unity) [reply Kernel/FileSystem/FATFileSystem.cpp:82]: Are you talking about the function?

File: Kernel/FileSystem/FATFileSystem.cpp (hunk 2 above)
Comment (Undefine-Unity) [reply Kernel/FileSystem/FATFileSystem.cpp:124]: What do you mean?

File: Kernel/FileSystem/FATFileSystem.cpp (hunk 1 above)
Comment (Quaker762) [reply Kernel/FileSystem/FATFileSystem.cpp:82]: No. The constant 2. This should be a named `constexpr`

File: Kernel/FileSystem/FATFileSystem.h [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+    u16 first_cluster_low;
+    u32 file_size;
+};
+static_assert(sizeof(FATEntry) == 32);
+
+struct [[gnu::packed]] FATLongFileNameEntry {
+    u8 entry_index;
+    u16 characters1[5];
+    u8 attribute;
+    u8 entry_type;
```
Comment (timschumi) [inline Kernel/FileSystem/FATFileSystem.h:70]: Is there an `enum` (or `enum`-like) type to go with this? (I do realize that the entry type seems to be currently unused by our code.)

File: Kernel/FileSystem/FATFileSystem.cpp [hunk 6]
```diff
@@ -0,0 +1,361 @@
...
+        filename.append(byte_terminated_string(m_entry.filename, 8, ' '));
+        if (m_entry.extension[0] != ' ') {
+            filename.append('.');
+            filename.append(byte_terminated_string(m_entry.extension, 3, ' '));
+        }
+        m_filename = MUST(KString::try_create(filename.string_view()));
+    } else {
+        StringBuilder filename;
+        for (auto& lfn_entry : lfn_entries) {
+
```
Comment (timschumi) [inline Kernel/FileSystem/FATFileSystem.cpp:225]: Empty line.

File: Kernel/FileSystem/FATFileSystem.cpp [hunk 7]
```diff
@@ -0,0 +1,361 @@
...
+    }
+
+    return EINVAL;
+}
+
+void FATInode::compute_filename(Vector<FATLongFileNameEntry> lfn_entries)
+{
+    VERIFY(!m_filename);
+
+    // FIXME: Find a way to handle errors here instead of using MUST
```
Comment (timschumi) [inline Kernel/FileSystem/FATFileSystem.cpp:213]: This function could probably already be `ErrorOr<void>`, or at least I don't see a reason why it couldn't be. Of course, this only moves the "handling errors" part further down the line, but it's a first step.

File: Kernel/Debug.h.in [hunk 8]
```diff
@@ -87,6 +87,14 @@
...
 #endif
 
+#ifndef FAT_DEBUG
+#cmakedefine01 FAT_DEBUG
+#endif
+
+#ifndef FAT_VERY_DEBUG
+#cmakedefine01 FAT_VERY_DEBUG
+#endif
+
```
Comment (timschumi) [inline Kernel/Debug.h.in:97]: Having two different log levels seems a bit over the top to me. The important parts of the initialization output (size, drive number, volume id, etc.) as well as some error messages could probably remain unguarded by a debug flag. The rest seems noisy enough (especially the lines on read/write operations) that it could be moved into `FAT_VERY_DEBUG` instead, which could then be the only debug flag, `FAT_DEBUG`.

File: Kernel/FileSystem/FATFileSystem.h [hunk 9]
```diff
@@ -0,0 +1,157 @@
...
+    virtual ~FATFS() override = default;
+    virtual ErrorOr<void> initialize() override;
+    virtual StringView class_name() const override { return "FATFS"sv; }
+    virtual Inode& root_inode() override;
+
+private:
+    FATFS(OpenFileDescription&);
+
+    static constexpr u8 signature_1 = 0x28;
+    static constexpr u8 signature_2 = 0x29;
```
Comment (timschumi) [inline Kernel/FileSystem/FATFileSystem.h:104]: Even though it's two bytes that aren't interpreted together, for things like signature checking we would usually just read and compare them as a single `u16`.

File: Kernel/FileSystem/FATFileSystem.h [hunk 10]
```diff
@@ -0,0 +1,157 @@
...
+
+    FAT32BootRecord const* boot_record() const { return (FAT32BootRecord const*)m_boot_record->data(); };
```
Comment (timschumi) [inline Kernel/FileSystem/FATFileSystem.h:106]: This should _probably_ be a C++ cast instead of a C cast, same with many other occurrences of casts in this file.

File: Kernel/FileSystem/FATFileSystem.cpp (hunk 4 above)
Comment (timschumi) [reply Kernel/FileSystem/FATFileSystem.cpp:317]: That sounds like an issue with `KString` though (or rather, where that specific `KString` is constructed) if unexpected null bytes are appearing, as `KString` does store a length.

File: Kernel/FileSystem/FATFileSystem.cpp [hunk 11]
```diff
@@ -0,0 +1,361 @@
...
+}
+
+// TODO: This might be worth putting as a part of the StringView class
+StringView FATInode::byte_terminated_string(u8 const* string, size_t max_length, u8 byte) const
+{
+    size_t length = max_length;
+
+    // HACK: size_t is unsigned, so we can't check if it's smaller than 0,
+    //       but when we go below zero the value turns really big
+    while (length <= max_length) {
```
Comment (timschumi) [inline Kernel/FileSystem/FATFileSystem.cpp:268]: That sounds like a prime opportunity for some weird values causing issues down the line (not that we probably survive until here anyways if something passes a length of `-1`).

Couldn't you start out with putting the `string` and `max_length` into a `StringView` at the very beginning, and then continue with cutting off bytes at the end? That would seem like a much safer alternative compared to working with raw pointers and offsets.

File: Kernel/FileSystem/FATFileSystem.cpp (hunk 1 above)
Comment (Undefine-Unity) [reply Kernel/FileSystem/FATFileSystem.cpp:82]: This really isn't a constant, it's just a part of the calculation.

File: Kernel/FileSystem/FATFileSystem.h [hunk 12]
```diff
@@ -0,0 +1,157 @@
...
+    u8 boot_jump[3];
+    u8 oem_identifier[8];
+    u16 bytes_per_sector;
+    u8 sectors_per_cluster;
+    u16 reserved_sector_count;
+    u8 fat_count;
+    u16 root_directory_entry_count;
+    u16 unused;
+    u8 media_descriptor_type;
+    u16 unused2;
```
[the remark this answers is not in the document]
Comment (Undefine-Unity) [reply Kernel/FileSystem/FATFileSystem.h]: There isn't any diffrence between `unused` and `reserved`, we aren't looking at any of these anyways. Changed the naming scheme to the same one used in the ISO9660 driver (start with `unused1` and increment).

File: Kernel/FileSystem/FATFileSystem.h [hunk 13]
```diff
@@ -0,0 +1,157 @@
...
+    u16 modification_date;
+    u16 first_cluster_low;
+    u32 file_size;
+};
+static_assert(sizeof(FATEntry) == 32);
+
+struct [[gnu::packed]] FATLongFileNameEntry {
+    u8 entry_index;
+    u16 characters1[5];
+    u8 attribute;
```
[the remark this answers is not in the document]
Comment (Undefine-Unity) [reply Kernel/FileSystem/FATFileSystem.h]: Long file name entries have their unique attribute and the value is always the same. I changed that to `attributes` just for consistency.

[the remark this answers is not in the document]
Comment (Undefine-Unity) [reply Kernel/FileSystem/FATFileSystem.h]: As I said below the value is always the same, but it is one of the possible attributes so I changed that.

File: Kernel/FileSystem/FATFileSystem.h (hunk 5 above)
Comment (Undefine-Unity) [reply Kernel/FileSystem/FATFileSystem.h:82]: The documentation says that it's always zero.

File: Kernel/FileSystem/FATFileSystem.cpp (hunk 7 above)
Comment (Undefine-Unity) [reply Kernel/FileSystem/FATFileSystem.cpp]: The only place that function is called is the constructor which is not failable. I made the function `ErrorOr<void>` and moved that `FIXME` to the constructor, but I don't know if that's a good solution.

File: Kernel/FileSystem/FATFileSystem.h (hunk 9 above)
Comment (Undefine-Unity) [reply Kernel/FileSystem/FATFileSystem.h:107]: They are not a one number, but two different possibilities.

File: Kernel/Debug.h.in (hunk 8 above)
Comment (Undefine-Unity) [reply Kernel/Debug.h.in]: While I agree that it's unnecessary to have two debug flags, but I still believe all the output should be guarded (except the invalid signature error).

File: Kernel/FileSystem/FATFileSystem.cpp [hunk 14]
```diff
@@ -0,0 +1,358 @@
...
+
+    return string;
+}
+
+ErrorOr<size_t> FATInode::read_bytes_locked(off_t offset, size_t size, UserOrKernelBuffer& buffer, OpenFileDescription*) const
+{
+    dbgln_if(FAT_DEBUG, "FATFS: Reading inode {}: ize: {} offset: {}", identifier().index(), size, offset);
+
+    // FIXME: Read only the needed blocks instead of the whole file
+    auto blocks = TRY(const_cast<FATInode&>(*this).read_block_list());
```
Comment (timschumi) [inline Kernel/FileSystem/FATFileSystem.cpp:280]: That looks very weird. Maybe move `generate_block_list` somewhere else and make `read_block_list` `const` as well.

**Review by timschumi — CHANGES_REQUESTED** [review]
Another round with what I originally missed.

File: Kernel/FileSystem/FATFileSystem.cpp [hunk 15]
```diff
@@ -0,0 +1,358 @@
...
+    time_t hour = time >> 11;
+    time_t minute = (time >> 5) & 0b1111111;
+    time_t second = (time & 0b11111) << 1;
+
+    return second + minute * 60 + hour * 3600
+        + day + (24 * 3600) + month * (30 * 24 * 3600) + year * (365 * 24 * 3600);
+}
+
+// TODO: This might be worth putting as a part of the StringView class
+StringView FATInode::byte_terminated_string(u8 const* initial_string, size_t max_length, u8 byte) const
```
[the remark this answers is not in the document]
Comment (Undefine-Unity) [reply Kernel/FileSystem/FATFileSystem.cpp]: It can't be made `char` because long file name entries are terminated by `0xFF` and the compiler would complain if we passed it as `char`.

File: Kernel/FileSystem/FATFileSystem.cpp (hunk 14 above)
Comment (Undefine-Unity) [reply Kernel/FileSystem/FATFileSystem.cpp:274]: There isn't an easy way to make `read_block_list` `const`. It uses some read functions that are not marked `const` (maybe they should be?) and also I think we should only compute the block list if we need in place like here, because let's say you want to list a directory with a very large amount of files, it would be pointless to run this function for every one of the files because we only about their file name.

File: Kernel/FileSystem/FATFileSystem.cpp (hunk 1 above)
Comment (Quaker762) [reply Kernel/FileSystem/FATFileSystem.cpp:82]: That's not true at all. The first two clusters are reserved special entries, hence `2` is something akin to `RESERVED_CLUSTER_COUNT`. This should be communicated instead of just sticking a random magic number into such a critical calculation.

File: Kernel/FileSystem/FATFileSystem.cpp [hunk 16]
```diff
@@ -0,0 +1,352 @@
...
+{
+    MutexLocker locker(m_inode_lock);
+
+    VERIFY(has_flag(m_entry.attributes, FATAttributes::Directory));
+
+    Vector<RefPtr<FATInode>> children = TRY(traverse());
+
+    for (auto& inode : children) {
+        if (strncmp(reinterpret_cast<char const*>(inode->m_filename->bytes().data()), reinterpret_cast<char const*>(name.bytes().data()), name.length()) == 0) {
+            return TRY(adopt_nonnull_lock_ref_or_enomem(inode.leak_ref()));
```
Comment (ADKaster) [inline Kernel/FileSystem/FATFileSystem.cpp:310]: This looks very suspicious.

Why are we leaking a reference? 

This makes it look like we have some places holding a RefPtr to these FATInode objects, and other places holding a Locking RefPtr. That seems very unsafe. If the parent INode classes expect things to be returned as locking RefPtrs, we should probably use that everywhere to not trip over ourselves later and fix the locking story at a higher level.

File: Kernel/FileSystem/FATFileSystem.cpp [hunk 17]
```diff
@@ -0,0 +1,352 @@
...
+ErrorOr<NonnullLockRefPtr<Inode>> FATInode::lookup(StringView name)
+{
+    MutexLocker locker(m_inode_lock);
+
+    VERIFY(has_flag(m_entry.attributes, FATAttributes::Directory));
+
+    Vector<RefPtr<FATInode>> children = TRY(traverse());
+
+    for (auto& inode : children) {
+        if (strncmp(reinterpret_cast<char const*>(inode->m_filename->bytes().data()), reinterpret_cast<char const*>(name.bytes().data()), name.length()) == 0) {
```
Comment (ADKaster) [inline Kernel/FileSystem/FATFileSystem.cpp:309]: This should just be `if (inode->m_filename->view() == name)`

It's very hard to read when you reach into both of these objects that are equality comparable to each other and reinterpret cast their bytes and use a C function.

File: Kernel/FileSystem/FATFileSystem.cpp [hunk 18]
```diff
@@ -0,0 +1,352 @@
...
+    return {};
+}
+
+ErrorOr<NonnullLockRefPtr<Inode>> FATInode::lookup(StringView name)
+{
+    MutexLocker locker(m_inode_lock);
+
+    VERIFY(has_flag(m_entry.attributes, FATAttributes::Directory));
+
+    Vector<RefPtr<FATInode>> children = TRY(traverse());
```
Comment (ADKaster) [inline Kernel/FileSystem/FATFileSystem.cpp:306]: I don't understand why we have to allocate a Vector of RefPtr to every child of this INode just to find the specific INode that has the passed-in name.

Can't traverse() take a `Function<bool(FATINode&)>` similar to other traverse/find functions in the system?

`RefPtr<FatINode> found_item = traverse_and_find([name](auto& child) {  return (child.m_filename.view() == name);' });`

File: Kernel/FileSystem/FATFileSystem.cpp [hunk 19]
```diff
@@ -0,0 +1,352 @@
...
+    for (u32 i = 0; i < blocks->size() / sizeof(FATEntry); i++) {
+        auto* entry = reinterpret_cast<FATEntry*>(blocks->data() + i * sizeof(FATEntry));
+        if (entry->filename[0] == 0x00) {
+            dbgln_if(FAT_DEBUG, "FATFS: Found end entry");
+            return inodes;
+        } else if (static_cast<u8>(entry->filename[0]) == 0xE5) {
+            dbgln_if(FAT_DEBUG, "FATFS: Found unused entry");
+        } else if (entry->attributes == FATAttributes::LongFileName) {
+            dbgln_if(FAT_DEBUG, "FATFS: Found LFN entry");
+            lfn_entries.append(*reinterpret_cast<FATLongFileNameEntry*>(entry));
```
Comment (ADKaster) [inline Kernel/FileSystem/FATFileSystem.cpp:196]: this can fail to allocate. as can the append on inodes below.

File: Kernel/FileSystem/FATFileSystem.cpp [hunk 20]
```diff
@@ -0,0 +1,352 @@
...
+    time_t second = (time & 0b11111) << 1;
+
+    return second + minute * 60 + hour * 3600
+        + day + (24 * 3600) + month * (30 * 24 * 3600) + year * (365 * 24 * 3600);
+}
+
+// TODO: This might be worth putting as a part of the StringView class
+StringView FATInode::byte_terminated_string(StringView string, u8 fill_byte)
+{
+    while (string.length() > 0) {
```
Comment (ADKaster) [inline Kernel/FileSystem/FATFileSystem.cpp:260]: It's very hard to figure out what this algorithm is trying to do.

It *looks* like you're doing some kind of `find_last_not_of`?

```
Optional<size_t> find_last_not_of(StringView haystack, char needle)
{
    for (size_t i = haystack.length(); i > 0; --i) {
        if (haystack[i - 1] != needle)
            return i - 1;
    }
    return {};
}


StringView FATInode::byte_terminated_string(StringView string, u8 fill_byte)
{
    if (auto index = find_last_not_of(string, fill_byte); index.has_value())
         return string.substring_view(0, index.value();
    return string;
}
```

File: Kernel/FileSystem/FATFileSystem.cpp [hunk 21]
```diff
@@ -0,0 +1,352 @@
...
+ErrorOr<void> FATInode::traverse_as_directory(Function<ErrorOr<void>(FileSystem::DirectoryEntryView const&)> callback) const
+{
+    MutexLocker locker(m_inode_lock);
+
+    VERIFY(has_flag(m_entry.attributes, FATAttributes::Directory));
+
+    Vector<RefPtr<FATInode>> children = TRY(const_cast<FATInode&>(*this).traverse());
+
+    for (auto const& inode : children) {
+        TRY(callback({ inode->m_filename->view(), inode->identifier(), static_cast<u8>(inode->m_entry.attributes) }));
```
Comment (ADKaster) [inline Kernel/FileSystem/FATFileSystem.cpp:294]: Same comment here.

We don't need to create a vector of RefPtrs just to invoke a callback on each item. we can call the callback while traversing.

File: Kernel/FileSystem/FATFileSystem.cpp [hunk 22]
```diff
@@ -0,0 +1,352 @@
...
+
+    if (cluster == boot_inode_index)
+        cluster = fs().boot_record()->root_directory_cluster;
+
+    Vector<BlockBasedFileSystem::BlockIndex> block_list;
+
+    auto fat_sector = TRY(KBuffer::try_create_with_size("FATFS: FAT read buffer"sv, fs().m_logical_block_size));
+    auto fat_sector_buffer = UserOrKernelBuffer::for_kernel_buffer(fat_sector->data());
+
+    while (cluster < 0x0FFFFFF8) {
```
Comment (ADKaster) [inline Kernel/FileSystem/FATFileSystem.cpp:132]: What do the magic numbers 0x0FFFFFF8 and 0x0FFFFFFF mean? Please give names to your constants. Similar to Quaker's comment on why the first two clusters are reserved.

If we have to read the FAT32 specification to understand the code, it is not descriptive enough.

**Review by ADKaster — COMMENTED** [review]
At the risk of adding too many cooks to the kitchen, here's my review...

I think the string iteration function has already had a few rounds of refactor, so feel free to ignore that if i'm talking nonsense.

I think that traversal is a bit unoptimized, but that can probably be addressed later.

The conversion of RefPtr to LockedRefPtr does look like a real thread safety  issue though...

File: Kernel/FileSystem/FATFileSystem.cpp (hunk 22 above)
Comment (bgianfo) [reply Kernel/FileSystem/FATFileSystem.cpp:134]: This seems like a valid critique of the code,  commenting constants isn't the serenity style. We name constants everywhere else in the codebase. This file system isn't special in that regard.

Please address this feedback.

Comment (Quaker762) [reply Kernel/FileSystem/FATFileSystem.cpp:134]: > I don't think that making constants for numbers used in only one place is a good idea. Instead I decided to add comments describing what do these numbers mean.

It significantly improves readability to someone that wants to poke around later on, and, as @bgianfo  has said, is in the Serenity style which dictates that there are to be no magic numbers, least of all not in the Kernel of all places.

Please fix this.

Comment (Undefine-Unity) [reply Kernel/FileSystem/FATFileSystem.cpp]: Done.

**Comment by Undefine-Unity** [comment]
I think I addressed all of the constants in one way or another. In the `fat_date_time` function I decided to add comments describing how times and dates are stored in FAT32, I think those comments describe what the code is doing in a way better way than having constants for all the masks and bitshifts. I replaced the rest of the magic numbers with constant variables.

File: Kernel/FileSystem/FATFileSystem.h [hunk 23]
```diff
@@ -0,0 +1,177 @@
...
+
+    static constexpr u32 no_more_clusters = 0x0FFFFFF8;
+    static constexpr u32 cluster_number_mask = 0x0FFFFFFF;
+
+    static constexpr u8 end_entry_byte = 0x00;
+    static constexpr u8 unused_entry_byte = 0xE5;
+
+    static constexpr u8 lfn_entry_text_termination = 0xFF;
+
+    static constexpr u16 first_fat_year_offset = 10; // 1980 (first possible year that can be stored as a fat date) - 1970 (first unix year)
```
Comment (timschumi) [inline Kernel/FileSystem/FATFileSystem.h:144]: Comment should be above the code that it comments, not behind it (unless the comment is _very_ short, which it isn't).

File: Kernel/FileSystem/FATFileSystem.cpp [hunk 24]
```diff
@@ -0,0 +1,357 @@
...
+
+    VERIFY_NOT_REACHED();
+}
+
+time_t FATInode::fat_date_time(u16 date, u16 time) const
+{
+    // Date format:
+    // - Year  7 bits
+    // - Month 4 bits
+    // - Day   5 bits
```
Comment (timschumi) [inline Kernel/FileSystem/FATFileSystem.cpp:249]: This information might be better stored in a proper bit-field. That makes the code easier to read, and it saves us from potential mistakes while shifting/masking. Same for the "Time format" below.

File: Kernel/FileSystem/FATFileSystem.cpp [hunk 25]
```diff
@@ -0,0 +1,357 @@
...
+            filename.append(lfn_entry.characters3[1]);
+        }
+
+        return TRY(KString::try_create(byte_terminated_string(filename.string_view(), lfn_entry_text_termination)));
+    }
+
+    VERIFY_NOT_REACHED();
+}
+
+time_t FATInode::fat_date_time(u16 date, u16 time) const
```
Comment (timschumi) [inline Kernel/FileSystem/FATFileSystem.cpp:244]: This should even be a `static` helper function, right? Not just `const`.

File: Kernel/FileSystem/FATFileSystem.cpp [hunk 26]
```diff
@@ -0,0 +1,357 @@
...
+    // Time format:
+    // - Hour 	5 bits
+    // - Minutes 6 bits
+    // - Seconds 5 bits (0 - 30, needs to be multiplied by two)
+
+    time_t hour = time >> 11;
+    time_t minute = (time >> 5) & 0b1111111;
+    time_t second = (time & 0b11111) * 2;
+
+    return second + minute * 60 + hour * 3600 + day + (24 * 3600) + month * (30 * 24 * 3600) + year * (365 * 24 * 3600);
```
Comment (timschumi) [inline Kernel/FileSystem/FATFileSystem.cpp:264]: Is assuming 30 days in a month and 365 days in a year correct?

File: Kernel/FileSystem/FATFileSystem.cpp [hunk 27]
```diff
@@ -0,0 +1,357 @@
...
+}
+
+time_t FATInode::fat_date_time(u16 date, u16 time) const
+{
+    // Date format:
+    // - Year  7 bits
+    // - Month 4 bits
+    // - Day   5 bits
+
+    time_t year = (date >> 9) + first_fat_year_offset;
```
Comment (timschumi) [inline Kernel/FileSystem/FATFileSystem.cpp:251]: In my opinion, this obstructs the clarity of the purpose of the offset quite a bit, even if it's appropriately explained with a comment where the constant is defined.

I'd rather do `time_t year = first_fat_year + (date >> 9);` here (where `first_fat_year` is 1980) and subtract the value 1970 in the formula at the end of the function (as that is the only part of the function that is actually concerned with making it a UNIX timestamp).

File: Kernel/FileSystem/FATFileSystem.cpp [hunk 28]
```diff
@@ -0,0 +1,357 @@
...
+    return m_metadata;
+}
+
+ErrorOr<void> FATInode::traverse_as_directory(Function<ErrorOr<void>(FileSystem::DirectoryEntryView const&)> callback) const
+{
+    MutexLocker locker(m_inode_lock);
+
+    VERIFY(has_flag(m_entry.attributes, FATAttributes::Directory));
+
+    [[maybe_unused]] auto inode = TRY(const_cast<FATInode&>(*this).traverse([&callback](auto inode) -> ErrorOr<bool> {
```
Comment (timschumi) [inline Kernel/FileSystem/FATFileSystem.cpp:296]: You could probably just drop the `[[maybe_unused]] auto inode = ` completely here.

**Review by timschumi — CHANGES_REQUESTED** [review]
Looks like it's coming together quite nicely. :^)

File: Kernel/FileSystem/FATFileSystem.cpp [hunk 29]
```diff
@@ -0,0 +1,357 @@
...
+    m_logical_block_size = boot_record()->bytes_per_sector;
+    set_block_size(m_logical_block_size);
+
+    u32 root_directory_sectors = ((boot_record()->root_directory_entry_count * sizeof(FATEntry)) + (m_logical_block_size - 1)) / m_logical_block_size;
+    m_first_data_sector = boot_record()->reserved_sector_count + (boot_record()->fat_count * boot_record()->sectors_per_fat) + root_directory_sectors;
+
+    TRY(BlockBasedFileSystem::initialize());
+
+    FATEntry root_entry {};
+    root_entry.first_cluster_low = FATInode::root_inode_index;
```
[the remark this answers is not in the document]
Comment (Undefine-Unity) [reply Kernel/FileSystem/FATFileSystem.cpp:69]: Yes they do, that's a part of the `FATEntry` struct which we do not have control of, because it's a part of the spec. I have added a helper function to read it.

File: Kernel/FileSystem/FATFileSystem.cpp (hunk 24 above)
Comment (Undefine-Unity) [reply Kernel/FileSystem/FATFileSystem.cpp:250]: Is there an pre-existing structure / class that I could use for that?

Comment (timschumi) [reply Kernel/FileSystem/FATFileSystem.cpp:250]: It would be something similar to [this](https://github.com/SerenityOS/serenity/blob/master/AK/FPControl.h#L41-L62), but obviously purpose-fit to the problem at hand.

I'd suggest the following (and a respective second definition for `FATPackedTime`):

```
union FATPackedDate {
    u16 value;
    struct {
        u16 year : 7;
        u16 month : 4;
        u16 day : 5; 
    };
};
static_assert(sizeof(FATPackedDate) == 16);
```

That way, you can access `packed_value.value` to read/write the raw value that the filesystem actually stores, and `packed_value.year`, `.month` and `.day` to read/write the parsed values.

This is also compatible with members inside structs, so you could define the various date and time members as that type directly (i.e. `FATPackedDate last_accessed_date` inside of `FATEntry`);

File: Kernel/FileSystem/FATFileSystem.cpp (hunk 28 above)
Comment (Undefine-Unity) [reply Kernel/FileSystem/FATFileSystem.cpp:302]: I can't, the compiler complains if I drop that.

File: Kernel/FileSystem/FATFileSystem.cpp (hunk 29 above)
[the remark this answers is not in the document]
Comment (Undefine-Unity) [reply Kernel/FileSystem/FATFileSystem.cpp]: I ended up getting rid of the `root_inode_index` constant and replaced it with the value that would be actually used later, now both `first_cluster_low` and `first_cluster_high` are being written when creating the root inode.

**Review by timschumi — CHANGES_REQUESTED** [review]
One last formality remains as far as I can tell, after that the PR gets a green check mark from me. :^)

**Review by timschumi — APPROVED** [review]
LGTM, and seems to work fine.
